File: ios/Approach/Sources/ModelsLibrary/Pin.swift
import Foundation

public enum Pin: Int, Equatable, Sendable, Identifiable, Codable, CaseIterable {
	case leftTwoPin = 0
	case leftThreePin = 1
	case headPin = 2
	case rightThreePin = 3
	case rightTwoPin = 4

	public var id: Int { rawValue }

	public var value: Int {
		switch self {
		case .leftTwoPin, .rightTwoPin: return 2
		case .leftThreePin, .rightThreePin: return 3
		case .headPin: return 5
		}
	}

	public static let fullDeck: Set<Self> = [.leftTwoPin, .leftThreePin, .headPin, .rightThreePin, .rightTwoPin]
}

extension Set where Element == Pin {
	public var value: Int { reduce(0) { value, pin in value + pin.value } }
	public var isHeadPin: Bool { count == 1 && first == .headPin }
	public var isHeadPin2: Bool { value == 7 && contains(.headPin) }
	public var isLeft: Bool { count == 4 && !contains(.leftTwoPin) }
	public var isRight: Bool { count == 4 && !contains(.rightTwoPin) }
	public var isAce: Bool { value == 11 }
	public var isLeftChopOff: Bool {
		count == 3 && contains(.headPin) && contains(.leftTwoPin) && contains(.leftThreePin)
	}
	public var isRightChopOff: Bool {
		count == 3 && contains(.headPin) && contains(.rightTwoPin) && contains(.rightThreePin)
	}
	public var isChopOff: Bool { isLeftChopOff || isRightChopOff }
	public var isLeftSplit: Bool { count == 2 && contains(.headPin) && contains(.leftThreePin) }
	public var isLeftSplitWithBonus: Bool {
		count == 3 && contains(.headPin) && contains(.leftThreePin) && contains(.rightTwoPin)
	}
	public var isRightSplit: Bool { count == 2 && contains(.headPin) && contains(.rightThreePin) }
	public var isRightSplitWithBonus: Bool {
		count == 3 && contains(.headPin) && contains(.rightThreePin) && contains(.leftTwoPin)
	}
	public var isSplit: Bool { isLeftSplit || isRightSplit }
	public var isSplitWithBonus: Bool { isLeftSplitWithBonus || isRightSplitWithBonus }
	public var isHitLeftOfMiddle: Bool { !contains(.headPin) && (contains(.leftTwoPin) || contains(.leftThreePin)) }
	public var isHitRightOfMiddle: Bool { !contains(.headPin) && (contains(.rightTwoPin) || contains(.rightThreePin)) }
	public var isMiddleHit: Bool { contains(.headPin) }
	public var isLeftTwelve: Bool { count == 4 && !contains(.rightThreePin) }
	public var isRightTwelve: Bool { count == 4 && !contains(.leftThreePin) }
	public var isTwelve: Bool { isLeftTwelve || isRightTwelve }
	public var arePinsCleared: Bool { count == 5 }
}

extension Set where Element == Pin {
	public func displayValue(rollIndex: Int) -> String {
		let outcome: RollOutcome
		if isHeadPin {
			outcome = .headPin
		} else if isHeadPin2 {
			outcome = .headPin2
		} else if isSplit {
			outcome = .split
		} else if isSplitWithBonus {
			outcome = .splitWithBonus
		} else if isChopOff {
			outcome = .chopOff
		} else if isAce {
			outcome = .ace
		} else if isLeft {
			outcome = .left
		} else if isRight {
			outcome = .right
		} else if arePinsCleared {
			switch rollIndex {
			case 0: outcome = .strike
			case 1: outcome = .spare
			default: outcome = .cleared
			}
		} else {
			outcome = .none
		}

		if outcome == .none {
			let value = self.value
			return value == 0 ? outcome.rawValue : String(value)
		} else {
			return rollIndex == 0 ? outcome.rawValue : outcome.numeral
		}
	}
}

public enum RollOutcome: String {
	case strike = "X"
	case spare = "/"
	case left = "L"
	case right = "R"
	case ace = "A"
	case chopOff = "C/O"
	case split = "HS"
	case splitWithBonus = "10"
	case headPin = "HP"
	case headPin2 = "H2"
	case cleared = "15"
	case none = "-"

	public var numeral: String {
		switch self {
		case .strike, .spare: return rawValue
		case .cleared: return "15"
		case .left, .right: return "13"
		case .ace: return "11"
		case .chopOff, .splitWithBonus: return "10"
		case .split: return "9"
		case .headPin2: return "7"
		case .headPin: return "5"
		case .none: return "-"
		}
	}
}

File: android/core/model/src/main/java/ca/josephroque/bowlingcompanion/core/model/Pin.kt
package ca.josephroque.bowlingcompanion.core.model

enum class Pin(val pinCount: Int) {
	LEFT_TWO_PIN(2),
	LEFT_THREE_PIN(3),
	HEAD_PIN(5),
	RIGHT_THREE_PIN(3),
	RIGHT_TWO_PIN(2);

	companion object {
		fun fullDeck(): Set<Pin> = entries.toSet()
	}
}

fun Set<Pin>.pinCount(): Int = sumOf(Pin::pinCount)

fun Set<Pin>.isHeadPin(): Boolean = size == 1 && contains(Pin.HEAD_PIN)
fun Set<Pin>.isHeadPin2(): Boolean = pinCount() == 7 && contains(Pin.HEAD_PIN)

fun Set<Pin>.isAce(): Boolean = pinCount() == 11

fun Set<Pin>.isLeftTap(): Boolean = size == 4 && !contains(Pin.LEFT_TWO_PIN)
fun Set<Pin>.isRightTap(): Boolean = size == 4 && !contains(Pin.RIGHT_TWO_PIN)
fun Set<Pin>.isTap(): Boolean = isLeftTap() || isRightTap()

fun Set<Pin>.isLeftChop(): Boolean = size == 3 && contains(Pin.HEAD_PIN) && contains(Pin.LEFT_TWO_PIN) && contains(Pin.LEFT_THREE_PIN)
fun Set<Pin>.isRightChop(): Boolean = size == 3 && contains(Pin.HEAD_PIN) && contains(Pin.RIGHT_TWO_PIN) && contains(Pin.RIGHT_THREE_PIN)
fun Set<Pin>.isChop(): Boolean = isLeftChop() || isRightChop()

fun Set<Pin>.isLeftSplit(): Boolean = size == 2 && contains(Pin.HEAD_PIN) && contains(Pin.LEFT_THREE_PIN)
fun Set<Pin>.isLeftSplitWithBonus(): Boolean = size == 3 && contains(Pin.HEAD_PIN) && contains(Pin.LEFT_THREE_PIN) && contains(Pin.RIGHT_TWO_PIN)
fun Set<Pin>.isRightSplit(): Boolean = size == 2 && contains(Pin.HEAD_PIN) && contains(Pin.RIGHT_THREE_PIN)
fun Set<Pin>.isRightSplitWithBonus(): Boolean = size == 3 && contains(Pin.HEAD_PIN) && contains(Pin.RIGHT_THREE_PIN) && contains(Pin.LEFT_TWO_PIN)
fun Set<Pin>.isSplit(): Boolean = isLeftSplit() || isRightSplit()
fun Set<Pin>.isSplitWithBonus(): Boolean = isLeftSplitWithBonus() || isRightSplitWithBonus()

fun Set<Pin>.isHitLeftOfMiddle(): Boolean = !contains(Pin.HEAD_PIN) && !contains(Pin.RIGHT_THREE_PIN) && (contains(Pin.LEFT_TWO_PIN) || contains(Pin.LEFT_THREE_PIN))
fun Set<Pin>.isHitRightOfMiddle(): Boolean = !contains(Pin.HEAD_PIN) && !contains(Pin.LEFT_THREE_PIN) && (contains(Pin.RIGHT_TWO_PIN) || contains(Pin.RIGHT_THREE_PIN))
fun Set<Pin>.isMiddleHit(): Boolean = contains(Pin.HEAD_PIN)

fun Set<Pin>.isLeftTwelve(): Boolean = size == 4 && !contains(Pin.RIGHT_THREE_PIN)
fun Set<Pin>.isRightTwelve(): Boolean = size == 4 && !contains(Pin.LEFT_THREE_PIN)
fun Set<Pin>.isTwelve(): Boolean = isLeftTwelve() || isRightTwelve()

fun Set<Pin>.isLeftFive(): Boolean = size == 2 && contains(Pin.LEFT_TWO_PIN) && contains(Pin.LEFT_THREE_PIN)
fun Set<Pin>.isRightFive(): Boolean = size == 2 && contains(Pin.RIGHT_TWO_PIN) && contains(Pin.RIGHT_THREE_PIN)
fun Set<Pin>.isFive(): Boolean = isLeftFive() || isRightFive()

fun Set<Pin>.isLeftThree(): Boolean = size == 1 && contains(Pin.LEFT_THREE_PIN)
fun Set<Pin>.isRightThree(): Boolean = size == 1 && contains(Pin.RIGHT_THREE_PIN)
fun Set<Pin>.isThree(): Boolean = isLeftThree() || isRightThree()

fun Set<Pin>.arePinsCleared(): Boolean = size == 5

fun Set<Pin>.displayAt(rollIndex: Int): String {
	val outcome: RollOutcome
	when {
		isHeadPin() -> outcome = RollOutcome.HEAD_PIN
		isHeadPin2() -> outcome = RollOutcome.HEAD_PIN_2
		isSplit() -> outcome = RollOutcome.SPLIT
		isSplitWithBonus() -> outcome = RollOutcome.SPLIT_WITH_BONUS
		isChop() -> outcome = RollOutcome.CHOP_OFF
		isAce() -> outcome = RollOutcome.ACE
		isLeftTap() -> outcome = RollOutcome.LEFT
		isRightTap() -> outcome = RollOutcome.RIGHT
		arePinsCleared() -> outcome = when (rollIndex) {
			0 -> RollOutcome.STRIKE
			1 -> RollOutcome.SPARE
			else -> RollOutcome.CLEARED
		}
		else -> outcome = RollOutcome.NONE
	}

	return if (outcome == RollOutcome.NONE) {
		val pinCount = this.pinCount()
		if (pinCount == 0) outcome.display else pinCount.toString()
	} else {
		if (rollIndex == 0) outcome.display else outcome.numericDisplay
	}
}

private enum class RollOutcome(val display: String, val numericDisplay: String) {
	STRIKE("X", "X"),
	SPARE("/", "/"),
	LEFT("L", "13"),
	RIGHT("R", "13"),
	ACE("A", "11"),
	CHOP_OFF("C/O", "10"),
	SPLIT("HS", "8"),
	SPLIT_WITH_BONUS("10", "10"),
	HEAD_PIN("HP", "5"),
	HEAD_PIN_2("H2", "7"),
	CLEARED("15", "15"),
	NONE("-", "-"),
}